File: src/image_yuv420.c
#include <caml/alloc.h>
#include <caml/bigarray.h>
#include <caml/memory.h>
#include <caml/misc.h>
#include <caml/mlvalues.h>
#include <caml/threads.h>

#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <limits.h>

#include "image_pixel.h"
#include "image_rgb.h"
#include "image_yuv420.h"

#define max(a,b) (a>b)?a:b
#define min(a,b) (a<b)?a:b
#define round(r,n) (((n+(r-1))/r)*r)

CAMLprim value caml_yuv420_fill(value img, value p)
{
  CAMLparam2(img, p);
  int y = Int_val(Field(p, 0));
  int u = Int_val(Field(p, 1));
  int v = Int_val(Field(p, 2));
  int height = YUV420_height(img);
  int y_stride = YUV420_y_stride(img);
  int uv_stride = YUV420_uv_stride(img);
  memset(YUV420_y(img), y, height*y_stride);
  memset(YUV420_u(img), u, round(2,height/2)*uv_stride);
  memset(YUV420_v(img), v, round(2,height/2)*uv_stride);
  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv420_to_int_image(value img)
{
  CAMLparam1(img);
  CAMLlocal2(ans,tmp);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i,j;
  int y,u,v,r,g,b,a;
  ans = caml_alloc_tuple(yuv.height);
  for (j = 0; j < yuv.height; j++)
    {
      tmp = caml_alloc_tuple(yuv.width);
      for (i = 0; i < yuv.width; i++)
        {
          y = Y(yuv,i,j);
          u = U(yuv,i,j);
          v = V(yuv,i,j);
          r = RofYUV(y,u,v);
          g = GofYUV(y,u,v);
          b = BofYUV(y,u,v);
          if (yuv.alpha)
            {
              a = A(yuv,i,j);
              r = r*a/0xff;
              g = g*a/0xff;
              b = b*a/0xff;
            }
          Store_field(tmp, i, Val_int((r<<16)+(g<<8)+b));
        }
      Store_field(ans, j, tmp);
    }
  CAMLreturn(ans);
}

CAMLprim value caml_yuv420_of_rgb24_string(value img, value s)
{
  CAMLparam2(img,s);
  yuv420 yuv;
  yuv420_of_value(&yuv,img);
  // We don't copy so we cannot release the lock
  unsigned char *data = (unsigned char *)String_val(s);
  int i,j;

  for (j = 0; j < yuv.height; j++)
    for (i = 0; i < yuv.width; i++)
      {
        int r = data[3*(j*yuv.width+i)+0];
        int g = data[3*(j*yuv.width+i)+1];
        int b = data[3*(j*yuv.width+i)+2];
        Y(yuv,i,j) = YofRGB(r,g,b);
        // TODO: don't do u/v twice
        U(yuv,i,j) = UofRGB(r,g,b);
        V(yuv,i,j) = VofRGB(r,g,b);
      }

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv420_of_rgba32(value _rgb, value img)
{
  CAMLparam2(_rgb,img);
   frame rgb;
  frame_of_value(_rgb, &rgb);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i, j;

  caml_enter_blocking_section();
  for (j = 0; j < rgb.height; j++)
    for (i = 0; i < rgb.width; i++)
      {
        int r = Red(&rgb,i,j);
        int g = Green(&rgb,i,j);
        int b = Blue(&rgb,i,j);
        Y(yuv,i,j) = YofRGB(r,g,b);
        // TODO: don't do u/v twice
        U(yuv,i,j) = UofRGB(r,g,b);
        V(yuv,i,j) = VofRGB(r,g,b);
        A(yuv,i,j) = Alpha(&rgb,i,j);
      }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv420_to_rgba32(value img, value _rgb)
{
  CAMLparam2(img,_rgb);
  frame rgb;
  frame_of_value(_rgb, &rgb);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i, j;

  caml_enter_blocking_section();
  for (j = 0; j < yuv.height; j++)
    for (i = 0; i < yuv.width; i++)
      {
        int y = Y(yuv,i,j);
        int u = U(yuv,i,j);
        int v = V(yuv,i,j);
        Red(&rgb,i,j) = RofYUV(y,u,v);
        Green(&rgb,i,j) = GofYUV(y,u,v);
        Blue(&rgb,i,j) = BofYUV(y,u,v);
        Alpha(&rgb,i,j) = yuv.alpha ? A(yuv,i,j) : 0xff;
      }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv420_scale(value _src, value _dst)
{
  CAMLparam2(_src, _dst);
  yuv420 src, dst;
  yuv420_of_value(&src, _src);
  yuv420_of_value(&dst, _dst);
  int i,j,is,js;

  assert(!src.alpha || dst.alpha);

  caml_enter_blocking_section();
  for (j = 0; j < dst.height; j++)
    for (i = 0; i < dst.width; i++)
      {
        is = i*src.width/dst.width;
        js = j*src.height/dst.height;
        Y(dst,i,j) = Y(src,is,js);
        // TODO: don't do u/v twice
        U(dst,i,j) = U(src,is,js);
        V(dst,i,j) = V(src,is,js);
        if (src.alpha) A(dst,i,j) = A(src,is,js);
      }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv420_scale_coef(value _src, value _dst, value xscale, value yscale)
{
  CAMLparam4(_src, _dst, xscale, yscale);
  yuv420 src, dst;
  yuv420_of_value(&src, _src);
  yuv420_of_value(&dst, _dst);
  // x scaling (xn: numerator, xd: denominator)
  int xn = Int_val(Field(xscale, 0));
  int xd = Int_val(Field(xscale, 1));
  // y scaling
  int yn = Int_val(Field(yscale, 0));
  int yd = Int_val(Field(yscale, 1));
  // offsets
  int ox = (dst.width - src.width * xn / xd) / 2;
  int oy = (dst.height - src.height * yn / yd) / 2;
  int i, j;

  assert(ox >= 0 && oy >= 0);

  caml_enter_blocking_section();
  // TODO: blank
  /* if (ox != 0 || oy != 0) rgb_blank(&dst); */
  for (j = oy; j < dst.height - oy; j++)
    for (i = ox; i < dst.width - ox; i++) {
      int is = (i-ox)*xd/xn;
      int js = (j-oy)*yd/yn;
      Y(dst,i,j) = Y(src,is,js);
      // TODO: don't do u/v twice
      U(dst,i,j) = U(src,is,js);
      V(dst,i,j) = V(src,is,js);
      if (src.alpha) A(dst,i,j) = A(src,is,js);
    }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}


CAMLprim value caml_yuv420_add(value _src, value _x, value _y, value _dst)
{
  CAMLparam4(_src, _x, _y, _dst);
  int x = Int_val(_x);
  int y = Int_val(_y);
  yuv420 src, dst;
  yuv420_of_value(&src, _src);
  yuv420_of_value(&dst, _dst);
  int ia = max(x,0);
  int ib = min(x+src.width,dst.width);
  int ja = max(y,0);
  int jb = min(y+src.height,dst.height);
  int i,j;

  caml_enter_blocking_section();
  if (src.alpha == NULL)
    for (j = ja; j < jb; j++)
      for (i = ia; i < ib; i++)
        {
          int is = i-x;
          int js = j-y;
          Y(dst,i,j) = Y(src,is,js);
          // TODO: don't do u/v twice
          U(dst,i,j) = U(src,is,js);
          V(dst,i,j) = V(src,is,js);
          if (dst.alpha) A(dst,i,j) = 0xff;
        }
  else
    for (j = ja; j < jb; j++)
      for (i = ia; i < ib; i++)
        {
          int is = i-x;
          int js = j-y;
          int a = A(src,is,js);

          if (a == 0) {}
          else if (a == 0xff)
            {
              Y(dst,i,j) = Y(src,is,js);
              U(dst,i,j) = U(src,is,js);
              V(dst,i,j) = V(src,is,js);
              if (dst.alpha) A(dst,i,j) = 0xff;
            }
          else
            {
              Y(dst,i,j) = CLIP((Y(src,is,js) * a + Y(dst,i,j) * (0xff - a)) / 0xff);
              // TODO: don't do u/v twice
              U(dst,i,j) = CLIP((U(src,is,js) * a + U(dst,i,j) * (0xff - a)) / 0xff);
              V(dst,i,j) = CLIP((V(src,is,js) * a + V(dst,i,j) * (0xff - a)) / 0xff);
              if (dst.alpha) A(dst,i,j) = 0xff-((0xff-a)*(0xff-A(dst,i,j)))/0xff;
            }
        }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv420_get_pixel_rgba(value img, value _i, value _j)
{
  CAMLparam3(img, _i, _j);
  CAMLlocal1(ans);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i = Int_val(_i);
  int j = Int_val(_j);

  int y = Y(yuv,i,j);
  int u = U(yuv,i,j);
  int v = V(yuv,i,j);
  int a = yuv.alpha ? A(yuv,i,j) : 0xff;
  int r = RofYUV(y,u,v);
  int g = GofYUV(y,u,v);
  int b = BofYUV(y,u,v);

  ans = caml_alloc_tuple(4);
  Store_field(ans, 0, Val_int(r));
  Store_field(ans, 1, Val_int(g));
  Store_field(ans, 2, Val_int(b));
  Store_field(ans, 3, Val_int(a));
  CAMLreturn(ans);
}

CAMLprim value caml_yuv420_set_pixel_rgba(value img, value _i, value _j, value c)
{
  CAMLparam4(img, _i, _j, c);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i = Int_val(_i);
  int j = Int_val(_j);
  int r = Int_val(Field(c,0));
  int g = Int_val(Field(c,1));
  int b = Int_val(Field(c,2));
  int a = Int_val(Field(c,3));

  Y(yuv,i,j) = YofRGB(r,g,b);
  U(yuv,i,j) = UofRGB(r,g,b);
  V(yuv,i,j) = VofRGB(r,g,b);
  if (yuv.alpha) A(yuv,i,j) = a;

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv_randomize(value img)
{
  CAMLparam1(img);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i, j;

  caml_enter_blocking_section();
  for (j = 0; j < yuv.height; j++)
    for (i = 0; i < yuv.width; i++) {
      Y(yuv,i,j) = rand();
      U(yuv,i,j) = rand();
      V(yuv,i,j) = rand();
    }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv_greyscale(value img)
{
  CAMLparam1(img);
  yuv420 yuv;
  yuv420_of_value(&yuv, img);
  int i, j;

  caml_enter_blocking_section();
  for (j = 0; j < yuv.height; j++)
    for (i = 0; i < yuv.width; i++) {
      /* int y = Y(yuv,i,j); */
      U(yuv,i,j) = 0x7f;
      V(yuv,i,j) = 0x7f;
    }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

#define PIXEL_PRECISON 0x10000
CAMLprim value caml_yuv_scale_alpha(value img, value _a)
{
  CAMLparam2(img, _a);
  yuv420 yuv;
  yuv420_of_value(&yuv,img);
  int a = Double_val(_a) * PIXEL_PRECISON;
  int i, j;

  caml_enter_blocking_section();
  for (j = 0; j < yuv.height; j++)
    for (i = 0; i < yuv.width; i++)
      A(yuv, i, j) = CLIP(A(yuv, i, j) * a / PIXEL_PRECISON);
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv_disk_alpha(value img, value _x, value _y, value _r)
{
  CAMLparam4(img, _x, _y, _r);
  yuv420 yuv;
  yuv420_of_value(&yuv,img);
  int x = Int_val(_x);
  int y = Int_val(_y);
  int radius = Int_val(_r);
  radius = radius * radius;
  int i, j;

  caml_enter_blocking_section();
  for (j = 0; j < yuv.height; j++)
    for (i = 0; i < yuv.width; i++)
    {
      int r = (i - x) * (i - x) + (j - y) * (j - y);
      if (r > radius) A(yuv, i, j) = 0;
    }
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

#define crop(x, m) (x > m ? m : (x < 0 ? 0 : x))

CAMLprim value caml_yuv_box_alpha_native(value img, value _x, value _y, value _w, value _h, value _a)
{
  CAMLparam1(img);
  yuv420 yuv;
  yuv420_of_value(&yuv,img);
  int x = crop(Int_val(_x), yuv.width);
  int y = crop(Int_val(_y), yuv.height);
  int w = crop(Int_val(_w), yuv.width);
  int h = max(Int_val(_h), yuv.height);
  int a = CLIP(Double_val(_a) * PIXEL_PRECISON);
  int i, j;

  caml_enter_blocking_section();
  for (j = y; j < h; j++)
    for (i = x; i < w; i++)
      A(yuv, i, j) = a;
  caml_leave_blocking_section();

  CAMLreturn(Val_unit);
}

CAMLprim value caml_yuv_box_alpha_bytecode(value * argv, int argn) {
  return caml_yuv_box_alpha_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
}

File: src/imageYUV420.ml
(*
 * Copyright 2011 The Savonet Team
 *
 * This file is part of ocaml-mm.
 *
 * ocaml-mm is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ocaml-mm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ocaml-mm; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * As a special exception to the GNU Library General Public License, you may
 * link, statically or dynamically, a "work that uses the Library" with a publicly
 * distributed version of the Library to produce an executable file containing
 * portions of the Library, and distribute that executable file under terms of
 * your choice, without any of the additional requirements listed in clause 6
 * of the GNU Library General Public License.
 * By "a publicly distributed version of the Library", we mean either the unmodified
 * Library as distributed by The Savonet Team, or a modified version of the Library that is
 * distributed under the conditions defined in clause 3 of the GNU Library General
 * Public License. This exception does not however invalidate any other reasons why
 * the executable file might be covered by the GNU Library General Public License.
 *
 *)

open ImageBase
module Bitmap = ImageBitmap
module RGBA32 = ImageRGBA32

type t = {
  mutable y : Data.t; (* This is a view over the whole data. *)
  mutable y_stride : int;
  mutable u : Data.t; (* This is a view over the whole data. *)
  mutable v : Data.t; (* This is a view over the whole data. *)
  mutable uv_stride : int;
  width : int;
  height : int;
  mutable alpha : Data.t option; (* This is a view over the whole data. *)
  mutable data : Data.t;
}

let width img = img.width
let height img = img.height
let uv_height height = Data.round 2 ((height + 1) / 2)
let dimensions img = (width img, height img)
let y img = img.y
let y_stride img = img.y_stride
let u img = img.u
let v img = img.v
let uv_stride img = img.uv_stride
let data img = img.data
let planes img = (img.y, img.u, img.v)
let alpha img = img.alpha
let set_alpha img alpha = img.alpha <- alpha
let size img = Data.size img.y + Data.size img.u + Data.size img.v

let ensure_alpha img =
  if img.alpha = None then (
    let a = Data.alloc (img.height * img.y_stride) in
    Data.fill a 0xff;
    img.alpha <- Some a)

external fill : t -> Pixel.yuv -> int -> unit = "caml_yuv420_fill"

let fill img pix = fill img pix (uv_height img.height)

let fill_alpha img a =
  if a = 0xff then img.alpha <- None
  else (
    ensure_alpha img;
    Bigarray.Array1.fill (Option.get img.alpha) a)

let blank img = fill img (Pixel.yuv_of_rgb (0, 0, 0))
let blank_all = blank

let sub_fields ~alpha ~height ~y_stride ~uv_stride data =
  let y = Data.sub data 0 (height * y_stride) in
  let u = Data.sub data (height * y_stride) (height / 2 * uv_stride) in
  let v =
    Data.sub data
      ((height * y_stride) + (height / 2 * uv_stride))
      (height / 2 * uv_stride)
  in
  let alpha =
    if alpha then
      Some
        (Data.sub data
           ((height * y_stride) + (height * uv_stride))
           (height * y_stride))
    else None
  in
  (y, u, v, alpha)

let data_len ~alpha ~y_stride ~uv_stride ~height () =
  height * (y_stride + uv_stride + if alpha then y_stride else 0)

let make_base ~alpha ~width ~height ~y_stride ~uv_stride () =
  let data = Data.alloc (data_len ~alpha ~y_stride ~uv_stride ~height ()) in
  let y, u, v, alpha = sub_fields ~alpha ~height ~y_stride ~uv_stride data in
  { data; y; y_stride; u; v; uv_stride; width; height; alpha }

let make width height ?alpha y y_stride u v uv_stride =
  let img =
    make_base ~alpha:(alpha <> None) ~width ~height ~y_stride ~uv_stride ()
  in
  Data.blit y 0 img.y 0 (height * y_stride);
  Data.blit u 0 img.u 0 (height / 2 * uv_stride);
  Data.blit v 0 img.v 0 (height / 2 * uv_stride);
  (match alpha with
    | None -> ()
    | Some a -> Data.blit a 0 (Option.get img.alpha) 0 (height * y_stride));
  img

let make_data ?(alpha = false) width height data y_stride uv_stride =
  assert (data_len ~alpha ~y_stride ~uv_stride ~height () <= Data.length data);
  let y, u, v, alpha = sub_fields ~alpha ~height ~y_stride ~uv_stride data in
  { data; y; y_stride; u; v; uv_stride; width; height; alpha }

(* Default alignment. *)
let align = Sys.word_size / 8

let default_stride width y_stride uv_stride =
  let y_stride = Option.value ~default:(Data.round align width) y_stride in
  let uv_stride =
    Option.value ~default:(Data.round align ((width + 1) / 2)) uv_stride
  in
  (y_stride, uv_stride)

let create ?(blank = false) ?y_stride ?uv_stride width height =
  let y_stride, uv_stride = default_stride width y_stride uv_stride in
  let y = Data.aligned align (height * y_stride) in
  let u, v =
    let height = uv_height height in
    ( Data.aligned align (height * uv_stride),
      Data.aligned align (height * uv_stride) )
  in
  let img = make width height y y_stride u v uv_stride in
  if blank then blank_all img;
  img

let has_alpha img = img.alpha <> None
let remove_alpha img = img.alpha <- None

let of_YUV420_string ?y_stride ?uv_stride s width height =
  (* let y_stride, uv_stride = default_stride width y_stride uv_stride in *)
  let y_stride = Option.value ~default:width y_stride in
  let uv_stride = Option.value ~default:(width / 2) uv_stride in
  let data = Data.of_string s in
  make_data width height data y_stride uv_stride

external of_RGB24_string : t -> string -> unit = "caml_yuv420_of_rgb24_string"

let of_RGB24_string s width =
  let height = String.length s / (3 * width) in
  let img = create width height in
  of_RGB24_string img s;
  img

external of_RGBA32 : RGBA32.t -> t -> unit = "caml_yuv420_of_rgba32"

let of_RGBA32 rgb =
  let width = RGBA32.width rgb in
  let height = RGBA32.height rgb in
  let img = create width height in
  ensure_alpha img;
  of_RGBA32 rgb img;
  img

external to_RGBA32 : t -> RGBA32.t -> unit = "caml_yuv420_to_rgba32"

let to_RGBA32 img =
  let width = img.width in
  let height = img.height in
  let rgb = RGBA32.create width height in
  to_RGBA32 img rgb;
  rgb

let of_PPM s =
  let img = of_RGBA32 (RGBA32.of_PPM s) in
  remove_alpha img;
  img

let to_BMP img =
  let img = to_RGBA32 img in
  RGBA32.to_BMP img

let copy img =
  let dst =
    create ~y_stride:img.y_stride ~uv_stride:img.uv_stride img.width img.height
  in
  Data.blit_all img.data dst.data;
  dst

let blit_all src dst =
  if src.width <> dst.width then raise Invalid_dimensions;
  if src.height <> dst.height then raise Invalid_dimensions;
  if Data.length src.y <> Data.length dst.y then (
    let tmp =
      make_base ~alpha:(src.alpha <> None) ~width:src.width ~height:src.width
        ~y_stride:src.y_stride ~uv_stride:src.uv_stride ()
    in
    dst.data <- tmp.data;
    dst.y <- tmp.y;
    dst.y_stride <- tmp.y_stride;
    dst.u <- tmp.u;
    dst.v <- tmp.v;
    dst.uv_stride <- tmp.uv_stride;
    dst.alpha <- tmp.alpha);
  Data.blit_all src.data dst.data

let blit src dst = blit_all src dst

external randomize : t -> unit = "caml_yuv_randomize"
external add : t -> int -> int -> t -> unit = "caml_yuv420_add"

let add src ?(x = 0) ?(y = 0) dst =
  (* Printf.printf "add %dx%d with %dx%d at %d,%d\n%!" (width src) (height src) (width dst) (height dst) x y; *)
  add src x y dst

external set_pixel_rgba : t -> int -> int -> Pixel.rgba -> unit
  = "caml_yuv420_set_pixel_rgba"

(* [@@noalloc] *)
let set_pixel_rgba img i j ((_, _, _, a) as p) =
  if not (0 <= i && i < img.width && 0 <= j && j < img.height) then
    raise Invalid_position;
  if a <> 0xff then ensure_alpha img;
  set_pixel_rgba img i j p

(*
  let set_pixel_rgba img i j (r,g,b,a) =
    let data = img.data in
    let width = img.width in
    let height = img.height in
    if img.alpha <> None || a <> 0xff then
      (
        ensure_alpha img;
        Bigarray.Array1.set (Option.get img.alpha) (j * width + i) a
      );
    let y,u,v = Pixel.yuv_of_rgb (r,g,b) in
    Bigarray.Array1.set data (j * width + i) y;
    Bigarray.Array1.set data (height * width + (j / 2) * (width / 2) + i / 2) u;
    Bigarray.Array1.set data (height * width * 5 / 4 + (j / 2) * (width / 2) + i / 2) v
   *)

let get_pixel_y img i j = Data.get img.y ((j * img.y_stride) + i)
let get_pixel_u img i j = Data.get img.u ((j / 2 * img.uv_stride) + (i / 2))
let get_pixel_v img i j = Data.get img.v ((j / 2 * img.uv_stride) + (i / 2))

let get_pixel_a img i j =
  match img.alpha with
    | Some alpha -> Data.get alpha ((j * img.y_stride) + i)
    | None -> 0xff

external get_pixel_rgba : t -> int -> int -> Pixel.rgba
  = "caml_yuv420_get_pixel_rgba"

let of_bitmap ?(fg = Pixel.RGBA.white) ?(bg = Pixel.RGBA.transparent) bmp =
  let width = Bitmap.width bmp in
  let height = Bitmap.height bmp in
  let img = create width height in
  for j = 0 to height - 1 do
    for i = 0 to width - 1 do
      set_pixel_rgba img i j (if Bitmap.get_pixel bmp i j then fg else bg)
    done
  done;
  img

external to_int_image : t -> int array array = "caml_yuv420_to_int_image"
external scale_full : t -> t -> unit = "caml_yuv420_scale"

let scale_full src dst =
  if has_alpha src then ensure_alpha dst;
  scale_full src dst

(** [scale_coef src dst (xn,xd) (yn,yd)] scales [src] into [dst] multiplying x
    dimension by xn/xd and y dimension by yn/yd. *)
external scale_coef : t -> t -> int * int -> int * int -> unit
  = "caml_yuv420_scale_coef"

let scale_proportional src dst =
  if has_alpha src then ensure_alpha dst;
  let sw, sh = (src.width, src.height) in
  let dw, dh = (dst.width, dst.height) in
  if dw = sw && dh = sh then blit_all src dst
  else (
    let n, d = if dh * sw < sh * dw then (dh, sh) else (dw, sw) in
    scale_coef src dst (n, d) (n, d))

let scale ?(proportional = false) src dst =
  if proportional then scale_proportional src dst else scale_full src dst

external rotate : t -> int -> int -> float -> t -> unit = "caml_yuv_rotate"

let rotate src x y a dst =
  ensure_alpha dst;
  rotate src x y a dst

external is_opaque : t -> bool = "caml_yuv_is_opaque"

let is_opaque img = if img.alpha = None then true else is_opaque img
let optimize_alpha img = if is_opaque img then img.alpha <- None

let alpha_to_y img =
  ensure_alpha img;
  img.y <- Option.get img.alpha;
  img.alpha <- None

external scale_alpha : t -> float -> unit = "caml_yuv_scale_alpha"

let scale_alpha img a =
  if a <> 1. then (
    ensure_alpha img;
    if a = 0. then fill_alpha img 0 else scale_alpha img a)

external disk_alpha : t -> int -> int -> int -> unit = "caml_yuv_disk_alpha"

let disk_alpha img x y r =
  ensure_alpha img;
  disk_alpha img x y r

external box_alpha : t -> int -> int -> int -> int -> float -> unit
  = "caml_yuv_box_alpha_bytecode" "caml_yuv_box_alpha_native"

let box_alpha img x y r =
  ensure_alpha img;
  box_alpha img x y r

external alpha_of_color : t -> int -> int -> int -> int -> unit
  = "caml_yuv_alpha_of_color"

let alpha_of_color img (y, u, v) d =
  ensure_alpha img;
  alpha_of_color img y u v d

external alpha_of_sameness : t -> t -> int -> unit
  = "caml_yuv_alpha_of_sameness"

let alpha_of_sameness ref img d =
  ensure_alpha img;
  alpha_of_sameness ref img d

external alpha_of_diff : t -> t -> int -> int -> unit = "caml_yuv_alpha_of_diff"

let alpha_of_diff ref img d s =
  ensure_alpha img;
  alpha_of_diff ref img d s

external gradient_uv : t -> int * int -> int * int -> int * int -> unit
  = "caml_yuv_gradient_uv"

external hmirror : t -> unit = "caml_yuv_hmirror"

module Effect = struct
  external greyscale : t -> unit = "caml_yuv_greyscale"
  external invert : t -> unit = "caml_yuv_invert"
  external sepia : t -> unit = "caml_yuv_sepia"
  external lomo : t -> unit = "caml_yuv_lomo"

  module Alpha = struct
    let scale = scale_alpha
    let disk = disk_alpha
  end
end
